gen-bpf.py

```diff
@@ -0,0 +1,314 @@
+#!/usr/bin/env python
+# usage: gen-bpf.py d_files_dir output_file
+
+from __future__ import print_function
+import re, ctypes, sys, json, binascii, os, time
+from pprint import pprint
+from collections import OrderedDict
+
+try:
+    (_prog, d_files_dir, output_file) = sys.argv
+except:
+    print("usage: %s h2o_path d_files_dir output_file" % sys.argv[0])
+    sys.exit(1)
+
+block_fields = {
+    "quicly:crypto_decrypt": set(["decrypted"]),
+    "quicly:receive": set(["bytes"]),
+    "quicly:crypto_update_secret": set(["secret"]),
+    "quicly:crypto_send_key_update": set(["secret"]),
+    "quicly:crypto_receive_key_update": set(["secret"]),
+    "quicly:crypto_receive_key_update_prepare": set(["secret"]),
+}
+
+block_probes = set([
+    "quicly:debug_message",
+])
+
+# mapping from proves.d's to quic-trace's:
+rename_map = {
+    "at": "time",
+    "master_id": "master_conn_id",
+}
+
+def read_from_file(path):
+    with open(path, "r") as f:
+        return f.read()
+
+def write_to_file(path, content):
+    with open(path, "w") as f:
+        f.write(content)
+
+d = read_from_file(os.path.join(d_files_dir, "quicly-probes.d"))
+
+whitespace = r'(?:/\*.*?\*/|\s+)'
+probe_decl = r'(?:\bprobe\s+(?:[a-zA-Z0-9_]+)\s*\([^\)]*\)\s*;)'
+d_decl = r'(?:\bprovider\s*(?P<provider>[a-zA-Z0-9_]+)\s*\{(?P<probes>(?:%s|%s)*)\})' % (probe_decl, whitespace)
+
+re_flags = re.X | re.M | re.S
+
+matched = re.search(d_decl, d, flags = re_flags)
+provider = matched.group('provider')
+
+struct_decl = read_from_file(os.path.join(os.path.dirname(__file__), "data-types.h"))
+
+def strip_typename(t):
+    return t.replace("*", "").replace("struct", "").replace("const", "").replace("strict", "").strip()
+
+def is_str_type(t):
+    return re.search(r'\b(?:char|u?int8_t|void)\s*\*', t)
+
+def is_ptr_type(t):
+    return "*" in t
+
+def is_bin_type(t):
+    return re.search(r'\b(?:u?int8_t|void)\s*\*', t)
+
+def build_tracer_name(metadata):
+    return "trace_%s__%s" % (metadata['provider'], metadata['name'])
+
+def build_tracer(metadata):
+    c = r"""
+int %s(struct pt_regs *ctx) {
+    void *buf = NULL;
+    struct event_t event = { .id = %d };
```

Comment:
Designated initializers(e.g. `.id = value`) is introduced since C++20 and almost all C++ compiler does not support C++20 completely. I suppose you should write this initialization as below

```c++
struct event_t event = { %d };
```
or


```c++
struct event_t event;
event_t.id = %d;
```

See also
- https://en.cppreference.com/w/cpp/language/aggregate_initialization

Reply:
Thanks. What's the hell!

Fortunately, however, that part is C built by BCC. Probabpy I should annotate it.

Reply:
I see